test/e2e/performanceprofile/functests/8_reboot/devices.go

```diff
@@ -0,0 +1,388 @@
+		if val, ok := os.LookupEnv(rebootCooldownEnvVar); ok {
+			rebootCooldownTime, err = time.ParseDuration(val)
+			Expect(err).ToNot(HaveOccurred(), "error creating workload pod")
+		}
+		testlog.Infof("reboot cooldown time: %v", rebootCooldownTime)
+
+		_, err = testnodes.GetByName(targetNode)
+		Expect(err).ToNot(HaveOccurred(), "error getting the target node %q", targetNode)
+	})
+
+	It("Verify that pods requesting devices are correctly recovered on node restart", func() {
+		namespace := testutils.NamespaceTesting
+
+		// refresh the targetNode object
+		node, err := testnodes.GetByName(targetNode)
+		Expect(err).ToNot(HaveOccurred(), "error getting the target node %q", targetNode)
+
+		// phase1: complete the node praparation: make sure we have enough devices. Short timeout, we should be idle now.
+		allocatableDevices := waitForNodeToReportResourcesOrFail("pre reboot", targetNode, sriovDeviceResourceName, 2*time.Minute, 2*time.Second)
+		Expect(allocatableDevices).To(BeNumerically(">=", minimumAllocatableDevices), "device %q has too low amount available - testing scenario unreliable", sriovDeviceResourceName, allocatableDevices)
+
+		// phase2: node is prepared, run the test workload and check it gets the device it expected
+		wlDp := makeWorkloadDeployment(namespace, workloadContainerImage, sriovDeviceResourceName, 1)
+		err = testclient.Client.Create(context.TODO(), wlDp)
+		Expect(err).ToNot(HaveOccurred(), "error creating workload deployment")
+
+		// short timeout: we are on idle cluster
+		wlDp = waitForDeploymentCompleteOrFail("pre reboot", *wlDp, 5*time.Minute)
+		testlog.Infof("deployment %s/%s running", wlDp.Namespace, wlDp.Name)
+
+		// phase3: the reboot, which trigger the very scenario we are after
+
+		// managed, clean restart (e.g. `reboot` command or `systemctl reboot`
+		// - details don't matter as long as this is a managed clean restart).
+		// Power loss scenarios, aka hard reboot, deferred to another test.
+		// intentionally ignoring error. We need to tolerate connection error or disconnect
+		// because the node is rebooting.
+		runCommandOnNodeThroughMCD(node, "reboot", rebootNodeCommandMCD)
+		// this is (likely) a SNO. We need to tolerate connection errors,
+		// because the apiserver is going down as well.
+		// we intentionally use a generous timeout.
+		// On Bare metal reboot can take a while.
+		waitForNodeReadyOrFail("post reboot", targetNode, 20*time.Minute, 3*time.Second)
+
+		// are we really sure? we can't predict if we will have state flapping,
+		// we can't predict if pods go back to containercreating and ideally we
+		// should have no flapping.
+		// Tracking all the state will make the test complex *and fragile*.
+		// The best we can do right now is to let the SNO cool down and check again.
+		testlog.Infof("post reboot: entering cooldown time: %v", rebootCooldownTime)
+		time.Sleep(rebootCooldownTime)
+		testlog.Infof("post reboot: finished cooldown time: %v", rebootCooldownTime)
```

Comment:
This is interesting, how did you come to the conclusion that a minute is sufficient for post reboot cooldown? Was it after some experimentation/testing with a real cluster?

Reply:
this is a tricky part. I did some experimentation and come up with "high enough" value. 
I'm trying to avoid the state flapping without adding fragile code, but I'm not so keen on sleeping.

Reply:
What I wasn't able to fully understand and am interested in learning here is that in spite of giving enough time (20 minutes) for the node to reboot why exactly did we need to wait another minute for the system to not be in flux? IOW, when you say "We expect things to be still on flux" what do we mean by that?

Reply:
The things I observed when SNO reboots is that  the state reported by apiserver flaps while the processes start up and expected communications resume. In practical terms, the apiserver SEEMS to answer queries with the latest consolidated status before reboot. For examples, pods are initially (= at the first query) reported Running, while it's not actually true, but the kubelet didn't update their status yet.

I must say I didn't dive too deep in here, but intuitively the theory has at least some merit.

This flapping may cause the test (or any other client) to misdetect things settled too early. The real pod status for example was observed going
Running -> Pending -> [a minute or so] -> Running again.

Tracking the state changes is hard and likely fragile, because on SNO reboot I didn't find a way to trust the apiserver until the cluster is actually settled, which is learned through a combination of factors (multiple parties reporting ready like node, all pods...).

Instead of chasing all these factors I added a sleep to let the cluster stabilize and bypass (some) flapping.
I think this can also be handled using a combination of `Eventually` and `Consistently`, but I didn't invest too much in this area yet.

Reply:
> The things I observed when SNO reboots is that the state reported by apiserver flaps while the processes start up and expected communications resume. In practical terms, the apiserver SEEMS to answer queries with the latest consolidated status before reboot. For examples, pods are initially (= at the first query) reported Running, while it's not actually true, but the kubelet didn't update their status yet.

I have noticed that as well and know exactly what you are talking about.
 
> This flapping may cause the test (or any other client) to misdetect things settled too early. The real pod status for example was observed going Running -> Pending -> [a minute or so] -> Running again.

What I noticed was that the recovered pod goes Running and then after the kubelet is back (in cases where device plugins are not up or haven't registered) it fails at  admission.

The new pod created is because of the deployment controller spinning up a new pod, goes into pending because the pod has to wait for the node allocatable to be updated and only when that happens the pod goes Running.

> Tracking the state changes is hard and likely fragile, because on SNO reboot I didn't find a way to trust the apiserver until the cluster is actually settled, which is learned through a combination of factors (multiple parties reporting ready like node, all pods...).
> 
> Instead of chasing all these factors I added a sleep to let the cluster stabilize and bypass (some) flapping. I think this can also be handled using a combination of `Eventually` and `Consistently`, but I didn't invest too much in this area yet.

I tend to be in the favor of using `Eventually` by waiting till the pod goes into admission error rather than putting an arbitrary sleep. I acknowledge that we could have cases where the pod correctly goes into running state when the SR-IOV devices are ready to be allocated. However, based on input from Sebastian and what I have noticed from my own testing it appears that if we have a large number of virtual function we can be fairly sure that it will take time for the config-daemon to bind them to the DPDK driver allowing us to be certain that the pod would fail at admission. Can we check for this in eventually as it is a more educated guess rather than an arbitrary sleep (which we cannot be sure about) ? 

Pretty much brainstorming here. Let me know what you think.

Reply:
> > This flapping may cause the test (or any other client) to misdetect things settled too early. The real pod status for example was observed going Running -> Pending -> [a minute or so] -> Running again.
> 
> What I noticed was that the recovered pod goes Running and then after the kubelet is back (in cases where device plugins are not up or haven't registered) it fails at admission.

This is exactly the problem: this pod is *first* wrongly reported Running, so `Eventually` won't help. We need `Eventually` + `Consistently`, both wrapped in a retry loop.

Pod goes running -> Eventually succeeds;
Pods STAYS running (is actually running for real) -> Consistently also succeeds

But if it's a flap, the Eventually succeeds and then the Consistently will fail (because the real state will put the pod back in Pending or even UnexpectedAdmissionError)
so we need to retry the pair a few times to really catch the correct sequence of events.

Doing like this looks more correct, but also looks more complex and fragile. Ignoring the state "for a while" seems a brutal yet surprisingly effective way to sidestep this issue.

> > Tracking the state changes is hard and likely fragile, because on SNO reboot I didn't find a way to trust the apiserver until the cluster is actually settled, which is learned through a combination of factors (multiple parties reporting ready like node, all pods...).
> > Instead of chasing all these factors I added a sleep to let the cluster stabilize and bypass (some) flapping. I think this can also be handled using a combination of `Eventually` and `Consistently`, but I didn't invest too much in this area yet.
> 
> I tend to be in the favor of using `Eventually` by waiting till the pod goes into admission error rather than putting an arbitrary sleep.

Note we do both: the sleep is "just" to skip the initial transient phase until components catch up.

>  I acknowledge that we could have cases where the pod correctly goes into running state when the SR-IOV devices are ready to be allocated. However, based on input from Sebastian and what I have noticed from my own testing it appears that if we have a large number of virtual function we can be fairly sure that it will take time for the config-daemon to bind them to the DPDK driver allowing us to be certain that the pod would fail at admission. Can we check for this in eventually as it is a more educated guess rather than an arbitrary sleep (which we cannot be sure about) ?

I agree we should improve here but I'm struggling to find a viable way forward :\

Reply:
Yep, this is a very tricky one indeed. Let's leave it as is for now and see if we can improve at a later stage.